Update documentation for SSL VERIFY hooks.

shinrich opened the pull request.

zwoop commented:
Hmmm, is this documentation for a yet to be landed hook? If so, why not merge this in with that commit?

zwoop commented:
Also, not that it matters much, but I think we (for some reason :-), try to keep documentation lines < 80 characters.

shinrich commented:
Both hooks already exist in the code.  Just wasn't documented yet.

shinrich commented:
I updated the docs and the constants to be consistent.  Both hooks already existing, but one was TS_SSL_VERIFY_CLIENT_HOOK and the other was TS_SSL_SERVER_VERIFY_HOOK.  I changed them both to the TS_SSL_VERIFY_* form.  Updated the enums, but left aliases of the old versions for backwards compatibility.

Will be adding tests in a separate PR.

CrendKing commented on doc/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.rst:
```diff
@@ -84,6 +84,30 @@ handshake processing will not proceed until :c:func:`TSSslVConnReenable()` is ca
+TS_SSL_VERIFY_SERVER_HOOK
+-------------------------
+
+This hooks is called when a client connects to Traffic Server and presents a
```
This comment does not sound correct. It seems to be copied verbatim from the TS_SSL_VERIFY_CLIENT_HOOK section.

CrendKing commented on doc/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.rst:
```diff
@@ -84,6 +84,30 @@ handshake processing will not proceed until :c:func:`TSSslVConnReenable()` is ca
+TS_SSL_VERIFY_CLIENT_HOOK
+-------------------------
+
+This hooks is called when a client connects to Traffic Server and presents a 
```
s/hooks/hook/

CrendKing commented on doc/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.rst:
```diff
@@ -84,6 +84,29 @@ handshake processing will not proceed until :c:func:`TSSslVConnReenable()` is ca
+TS_SSL_VERIFY_SERVER_HOOK
+-------------------------
+
+This hooks is called when a Traffic Server connections to an origin and the origin
```
s/connections/connects/

shinrich commented:
Updated to address @CrendKing's comment.

shinrich commented:
@CrendKing @zwoop Any other comments or are we good to go?

shinrich commented on doc/developer-guide/plugins/hooks-and-transactions/ssl-hooks.en.rst:
```diff
@@ -84,6 +84,29 @@ handshake processing will not proceed until :c:func:`TSSslVConnReenable()` is ca
+presents a certificate.  The callback can get the SSL object from the TSVConn
+argument and use that to access the origin certificate and make any additional checks.
+
+Processing will continue regardless of whether the hook callback executes
```
Correct.  The underlying openssl structure limits what we can do.  However, in the case of the verify_server hook, we cannot leave the callback not-enabled.  But we can return an ERROR and have the core ATS verify_callback interpret that to fail the certificate verification.

CrendKing replied:
>  The underlying openssl structure limits what we can do

Just to make sure we are on the side page. We agree that if the callback function from `SSL_set_verify()` returns 0, the OpenSSL will stop the TLS handshake, right? So what do you mean here?

> The return value of verify_callback controls the strategy of the further verification process. If verify_callback returns 0, the verification process is immediately stopped with "verification failed" state

shinrich replied:
Correct.  The verify_callback() function is registered by Traffic Server via the SSL_CTX_set_verify or SSL_set_verify openssl function.  It will eventually return either 0 or 1.  If it returns 0, the handshake will fail . It it returns 1, the handshake will continue on.

The VERIFY_SERVER_HOOK gives a place for plugins to insert logic and influence the return value of the verify_callback().  If the hook reenables with a TS_EVENT_ERROR, the verify_callback() will return 0.

bryancall commented:
Because of conflicts a new PR was created to merge into 8.1.0: #5236
